### PR TITLE
qs - update types to allow for custom type when parsing

### PR DESCRIPTION
**Pull request by ghost**
The change I've added will allow for

`parse<SomeType>(search, { ignoreQueryPrefix: true });`

Currently I have to do `parse(search, { ignoreQueryPrefix: true }); as SomeType`

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: Not the same package, but the same approach is talked about here <<https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30205>>

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## qs ([<kbd>unpkg</kbd>](https://unpkg.com/browse/qs@latest/))
was missing the following properties:
1. formats
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4f855088d5f3b1233744e1d8060d73c040f65551
</p>

File: types/qs/index.d.ts [hunk 1]
```diff
@@ -54,9 +54,11 @@ declare namespace QueryString {
...
 
-    interface ParsedQs { [key: string]: undefined | string | string[] | ParsedQs | ParsedQs[] }
+    interface ParsedQs {
+        [key: string]: undefined | string | string[] | ParsedQs | ParsedQs[];
+    }
 
     function stringify(obj: any, options?: IStringifyOptions): string;
-    function parse(str: string, options?: IParseOptions & { decoder?: never }): ParsedQs;
-    function parse(str: string, options?: IParseOptions): { [key: string]: unknown };
+    function parse<T = ParsedQs>(str: string, options?: IParseOptions & { decoder?: never }): T;
```
Comment (ljharb) [inline types/qs/index.d.ts:62]: Should this be `T extends ParsedQs`?

**Review by ljharb — CHANGES_REQUESTED** [review]

**Comment by jablko** [comment]
Does this advice from the README apply to this change? https://github.com/DefinitelyTyped/DefinitelyTyped#user-content-common-mistakes:~:text=If%20a%20type%20parameter%20does%20not,just%20have%20a%20disguised%20type%20assertion.

**Comment by Ed-Ed** [comment]
> Does this advice from the README apply to this change? https://github.com/DefinitelyTyped/DefinitelyTyped#user-content-common-mistakes:~:text=If%20a%20type%20parameter%20does%20not,just%20have%20a%20disguised%20type%20assertion.

Yeah that makes sense, I must have missed that my bad, will delete the PR.

**Comment by tim-phillips** [comment]
Copying the text here since the above link is broken:

> getMeAT<T>(): T: **If a type parameter does not appear in the types of any parameters, you don't really have a generic function, you just have a disguised type assertion.** Prefer to use a real type assertion, e.g. getMeAT() as number. Example where a type parameter is acceptable: function id<T>(value: T): T;. Example where it is not acceptable: function parseJson<T>(json: string): T;. Exception: new Map<string, number>() is OK.

https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes

**Pull request closed without merging**